### ISSUE TITLE
Test failing on Mac with perl 5.22.0

### ISSUE THREAD

**Issue by mmucklo** [msgpack/msgpack-perl#23]
Fresh compile of perl 5.22.0 on OS 10.11 (El Capitan)

I keep getting the following test failing (I've tried msgpack 0.39 through 0.48):

```
t/05_preferred_int.t ........ 1/22 
#   Failed test 'dump d2 80 00 00 00'
#   at t/05_preferred_int.t line 18.
#          got: 'ce 80 00 00 00'
#     expected: 'd2 80 00 00 00'
# Looks like you failed 1 test of 22.
```

Here's a summary of my perl5 environment:

```
# local/perl5/bin/perl -Ilib -V
Summary of my perl5 (revision 5 version 22 subversion 0) configuration:
   
  Platform:
    osname=darwin, osvers=15.0.0, archname=darwin-2level
    uname='darwin mmucklo.somewhere 15.0.0 darwin kernel version 15.0.0: wed aug 26 16:57:32 pdt 2015; root:xnu-3247.1.106~1release_x86_64 x86_64 '
    config_args='-des -Dprefix=/Users/mmucklo/somewhere/local/perl5 -Dlibpth=/Users/mmucklo/somewhere/local/lib -Dldflags=-L/Users/mmucklo/somewhere/local/lib -Dccflags=-I/Users/mmucklo/somewhere/local/include'
    hint=recommended, useposix=true, d_sigaction=define
    useithreads=undef, usemultiplicity=undef
    use64bitint=define, use64bitall=define, uselongdouble=undef
    usemymalloc=n, bincompat5005=undef
  Compiler:
    cc='cc', ccflags ='-I/Users/mmucklo/somewhere/local/include -fno-common -DPERL_DARWIN -mmacosx-version-min=10.11 -fno-strict-aliasing -pipe -fstack-protector-strong -I/usr/local/include',
    optimize='-O3',
    cppflags='-I/Users/mmucklo/somewhere/local/include -fno-common -DPERL_DARWIN -mmacosx-version-min=10.11 -fno-strict-aliasing -pipe -fstack-protector-strong -I/usr/local/include'
    ccversion='', gccversion='4.2.1 Compatible Apple LLVM 7.0.0 (clang-700.1.76)', gccosandvers=''
    intsize=4, longsize=8, ptrsize=8, doublesize=8, byteorder=12345678, doublekind=3
    d_longlong=define, longlongsize=8, d_longdbl=define, longdblsize=16, longdblkind=3
    ivtype='long', ivsize=8, nvtype='double', nvsize=8, Off_t='off_t', lseeksize=8
    alignbytes=8, prototype=define
  Linker and Libraries:
    ld='cc', ldflags ='-L/Users/mmucklo/somewhere/local/lib -mmacosx-version-min=10.11 -fstack-protector-strong -L/usr/local/lib'
    libpth=/Users/mmucklo/somewhere/local/lib /usr/local/lib /Applications/Xcode.app/Contents/Developer/Toolchains/XcodeDefault.xctoolchain/usr/bin/../lib/clang/7.0.0/lib /Applications/Xcode.app/Contents/Developer/Toolchains/XcodeDefault.xctoolchain/usr/lib /usr/lib
    libs=-lpthread -ldbm -ldb -ldl -lm -lutil -lc
    perllibs=-lpthread -ldl -lm -lutil -lc
    libc=, so=dylib, useshrplib=false, libperl=libperl.a
    gnulibc_version=''
  Dynamic Linking:
    dlsrc=dl_dlopen.xs, dlext=bundle, d_dlsymun=undef, ccdlflags=' '
    cccdlflags=' ', lddlflags='-L/Users/mmucklo/somewhere/local/lib -mmacosx-version-min=10.11 -bundle -undefined dynamic_lookup -L/usr/local/lib -fstack-protector-strong'


Characteristics of this binary (from libperl): 
  Compile-time options: HAS_TIMES PERLIO_LAYERS PERL_DONT_CREATE_GVSV
                        PERL_HASH_FUNC_ONE_AT_A_TIME_HARD PERL_MALLOC_WRAP
                        PERL_NEW_COPY_ON_WRITE PERL_PRESERVE_IVUV
                        USE_64_BIT_ALL USE_64_BIT_INT USE_LARGE_FILES
                        USE_LOCALE USE_LOCALE_COLLATE USE_LOCALE_CTYPE
                        USE_LOCALE_NUMERIC USE_LOCALE_TIME USE_PERLIO
                        USE_PERL_ATOF
  Built under darwin
  Compiled at Oct 28 2015 18:25:48
  @INC:
    lib
    /Users/mmucklo/somewhere/local/perl5/lib/site_perl/5.22.0/darwin-2level
    /Users/mmucklo/somewhere/local/perl5/lib/site_perl/5.22.0
    /Users/mmucklo/somewhere/local/perl5/lib/5.22.0/darwin-2level
    /Users/mmucklo/somewhere/local/perl5/lib/5.22.0
    .
```

**Comment by syohex** [comment]
I can reproduce this issue. However I cannot reproduce with lower optimization option(`-O0`, `-O1`). Could you check test by following steps ?

- `make distclean`
- `perl  Makefile.PL`
- Edit `Makefile` and replace `-O3` option with `-O0` or `-O1`
- `make`
- Test again `prove -bv t/05_preferred_int.t`

I suppose this may be compiler issue.

**Comment by mmucklo** [comment]
-O1 does indeed work.

**Comment by kazuho** [comment]
Not being able to reproduce I am not sure, but my two cents go to this line: [pack.c line 161](https://github.com/msgpack/msgpack-perl/blob/master/xs-src/pack.c#L161)

> ```
> msgpack_pack_int32(enc, ((int32_t)num) * -1);
```

In case num is `0x80000000`, the multiplication overflows, which leads to undefined behavior.

If this is the case, changing the expression to `(int32_t)-(int64_t)num` will likely fix the issue.

**Comment by tony** [comment]
Changing it from `-O2` to `-O1` in `Makefile` resolved the issue on FreeBSD 11-CURRENT

```
$ freebsd-version; uname -a                                          
11.0-CURRENT
FreeBSD z600 11.0-CURRENT FreeBSD 11.0-CURRENT #14: Sat Nov 14 00:58:49 CST 2015 

$ perl --version

This is perl 5, version 20, subversion 3 (v5.20.3) built for amd64-freebsd-thread-multi

$ clang --version
FreeBSD clang version 3.7.0 (tags/RELEASE_370/final 246257) 20150906
Target: x86_64-unknown-freebsd11.0
Thread model: posix
```

**Comment by kazuho** [comment]
@tony @mmucklo Can you please check if applying #24 fixes the issue?

**Comment by mmucklo** [comment]
Yes, applying #24 fixed the issue for me.

**Comment by tony** [comment]
@kazuho That did it. Tests pass on FreeBSD 11-CURRENT with Clang 3.7.

**Comment by kazuho** [comment]
@mmucklo @tony Thank you for the confirmation.

@gfx @tokuhirom @makamaka please respond to @shohex on #24.

**Comment by syohex** [comment]
Merged #24. Thanks @kazuho

**Comment by syohex** [comment]
I have released version 0.49(https://metacpan.org/release/Data-MessagePack) and confirmed that tests are passed on MacOSX 10.11, Clang(Apple LLVM version 7.0.0 (clang-700.1.76))
